src/java.base/share/classes/java/lang/Throwable.java

```diff
@@ -687,41 +687,59 @@ public void printStackTrace(PrintStream s) {
     }
 
     private void printStackTrace(PrintStreamOrWriter s) {
-        // Guard against malicious overrides of Throwable.equals by
-        // using a Set with identity equality semantics.
-        Set<Throwable> dejaVu = Collections.newSetFromMap(new IdentityHashMap<>());
-        dejaVu.add(this);
+        /*
+         * Guard against malicious overrides of Throwable.equals by
+         * using a Set with identity equality semantics.
+         */
+        Set<Throwable> dejaVu = null;
 
         synchronized(s.lock()) {
             // Print our stack trace
             s.println(this);
             StackTraceElement[] trace = getOurStackTrace();
-            for (StackTraceElement traceElement : trace)
-                s.println("\tat " + traceElement);
+            printStackTrace0(s, trace);
 
             // Print suppressed exceptions, if any
-            for (Throwable se : getSuppressed())
-                se.printEnclosedStackTrace(s, trace, SUPPRESSED_CAPTION, "\t", dejaVu);
+            java.lang.Throwable[] suppressed = getSuppressed();
+            if (suppressed.length > 0) {
+                dejaVu = dejaVu();
```

Comment:
Instead of creating dejaVu() here and there, I think you can do after printStackTrace0:
```
printStackTrace0(s, trace);

var suppressed = getSuppressed();
var ourCause = getCause();

if (suppressed.length > 0 || cause != null) {
    // Guard against malicious overrides of Throwable.equals by
    // using a Set with identity equality semantics.
    Set<Throwable> dejaVu = Collections.newSetFromMap(new IdentityHashMap<>());
    dejaVu.add(this);
    // more code here
}
```